app/features/worksheet/tools.py

```diff
@@ -0,0 +1,311 @@
+            template=self.prompt_template,
+            input_variables=["topic", "grade_level"],
+            partial_variables={"format_instructions": self.parser.get_format_instructions()}
+        )
+        
+        self.chain = prompt | self.model | self.parser
+        
+        if self.verbose: logger.info(f"Chain for {self.section_name} compilation complete")
+    
+    def validate_response(self) -> bool:
+        pass
+
+    def not_unique(self):
+        question = self.response[self.main_field]
+        question_embedding = self.embedding_model.encode(question, convert_to_tensor=True)
+
+        for q in self.question_bank:
+            similarity = util.pytorch_cos_sim(question_embedding, q)
+            if similarity.item() > self.similarity_threshold:
+                return True
+
+        self.question_bank.append(question_embedding)
+        return False
+    
+    def create_questions(self, num_questions = 0) -> List:
+        attempts = 0
+        max_attempts = num_questions * 5  # Allow for more attempts to generate questions
+        self.question_bank = []
+        generated_questions = []
+
+        while len(generated_questions) < num_questions and attempts < max_attempts:
+            # Move to the next attempt regardless of success to ensure progress
+            attempts += 1
+            self.response = self.chain.invoke({"topic": self.topic, "grade_level": self.grade_level})
+
+            if self.verbose:
+                logger.info(f"Generated {self.section_name} response attempt {attempts}: {self.response}")
+            
+            # Directly check if the response format is valid
+            flag = self.validate_response()
+            if flag:
+                # check if question is unique
+                if self.section_name != 'Summary' and self.not_unique():
+                    logger.warning(f"Not unique {self.section_name}. Attempt {attempts} of {max_attempts}")
+                    continue
+                generated_questions.append(self.response)
+                if self.verbose:
+                    logger.info(f"Valid {self.section_name} added: {self.response}")
+                    logger.info(f"Total generated {self.section_name}(s): {len(generated_questions)}")
+            else:
+                if self.verbose:
+                    logger.warning(f"Invalid {self.section_name} response format. Attempt {attempts} of {max_attempts}")
+        
+        # Log if fewer questions are generated
+        if len(generated_questions) < num_questions:
+            logger.warning(f"Only generated {len(generated_questions)} out of {num_questions} requested {self.section_name}(s)")
+
+        return generated_questions
+
+# Multiple Choice Question subclass
+class MultipleChoiceQuestion(QuestionBase):
+    def set_names(self):
+        self.section_name = 'Miltiple-Choice question'
+        self.main_field = 'question'
+
+    def validate_response(self) -> bool:
+        try:
```

Comment:
Let's refactor this

Reply:
Research on Pydantic on validations

Reply:
https://refactoring.guru/design-patterns/factory-method

Either pydantic or since your use OOP, try using a Factory Design Pattern

Reply:
Fixed it. Used pydantic ValidationError.